include/openthread/border_router.h

```diff
@@ -168,6 +168,54 @@ otError otBorderRouterGetNextRoute(otInstance *           aInstance,
  */
 otError otBorderRouterRegister(otInstance *aInstance);
 
+/**
+ * This function sends a datagram via the Thread interface with the logic (like NAT64) for border routers.
+ *
+ * The caller transfers ownership of @p aMessage when making this call. OpenThread will free @p aMessage when processing
+ * is complete, including when a value other than `OT_ERROR_NONE` is returned.
+ *
+ * The content of @p aMessage can be an IPv6 packet or an IPv4 packet. When @p aMessage contains an IPv4 packet, NAT64
+ * will work and translate it into an IPv6 packet. @p aMessage should be allocated by otIp4NewMessage if the
+ * content is an IPv4 packet.
+ *
+ * @param[in]  aInstance A pointer to an OpenThread instance.
+ * @param[in]  aMessage  A pointer to the message buffer containing the IPv6 datagram.
+ *
+ * @retval OT_ERROR_NONE                    Successfully processed the message.
+ * @retval OT_ERROR_DROP                    Message was well-formed but not fully processed due to packet processing
+ * rules.
+ * @retval OT_ERROR_NO_BUFS                 Could not allocate necessary message buffers when processing the datagram.
+ * @retval OT_ERROR_NO_ROUTE                No route to host.
+ * @retval OT_ERROR_INVALID_SOURCE_ADDRESS  Source address is invalid, e.g. an anycast address or a multicast address.
+ * @retval OT_ERROR_PARSE                   Encountered a malformed header when processing the message.
+ *
+ * @sa otIp4NewMessage
+ * @sa otBorderRouterSetReceiveCallback
+ *
+ */
+otError otBorderRouterSend(otInstance *aInstance, otMessage *aMessage);
+
+/**
+ * This function registers a callback to provide received datagrams for border routers.
+ *
+ * By default, this callback does not pass Thread control traffic.  See otIp6SetReceiveFilterEnabled() to
+ * change the Thread control traffic filter setting.
+ *
+ * The callback registered by this function may receive an IPv4 packet when NAT64 is enabled.
+ *
+ * @param[in]  aInstance         A pointer to an OpenThread instance.
+ * @param[in]  aCallback         A pointer to a function that is called when an IPv6 datagram is received or
+ *                               NULL to disable the callback.
+ * @param[in]  aCallbackContext  A pointer to application-specific context.
+ *
+ * @sa otBorderRouterSend
+ * @sa otIp6SetReceiveCallback
+ * @sa otIp6IsReceiveFilterEnabled
+ * @sa otIp6SetReceiveFilterEnabled
+ *
+ */
+void otBorderRouterSetReceiveCallback(otInstance *aInstance, otIp6ReceiveCallback aCallback, void *aCallbackContext);
```

Comment:
We are adding new public APIs `otBorderRouterSend` and `otBorderRouterSetReceiveCallback` for sending and registering a callback to receive IPv6 or IPv4 messages (which then uses NAT64).

This is actually replacing the existing `ip6.h` APIs `otIp6Send()` and `otIp6SetReceiveCallback()`. 

We basically require that on BR, the user of OT APIs, to switch to using new APIs instead of the older `otIp6` ones (even for IPv6 traffic). In particular, related to receive callback we can have confusion situation:
  - User can register a callback using `otIp6SetReceiveCallback()` which then expects to receive IPv6 message.
  - But if it uses `otBorderRouterSetReceiveCallback()` as well it replaces the other callback.

Reply:
`otNat64SendIp4` which only sends IPv4 packets, and it will work,

However for`otNat64SetReceiveIp4Callback`  The NAT64 prefix is recorded in the routing manager and Nat64 translator, which means the Ip6 module won't know whether a packet should be translated into IPv4 unless it routes it to the routing manager or the NAT64 translator.

Infact, for the BorderRouter itself, it really don't have to worry about whether the datagram is an IPv4 or an IPv6 message (or just peek the first byte if they want to do so)

Reply:
> However for `otNat64SetReceiveIp4Callback` The NAT64 prefix is recorded in the routing manager and Nat64 translator, which means the Ip6 module won't know whether a packet should be translated into IPv4 unless it routes it to the routing manager or the NAT64 translator.

Yes. That should be possible. In `Ip6::ProcessReceiveCallback()` (which does pass the message to host/user calling the registered `mReceiveIp6DatagramCallback`) we can call a `Nat64Translator` method and pass in `Message` to it for `Nat64Translator` to have a first pass at the message.